src/DataManagement/MonthlyReport/MonthlyReportCalculator.cs

```diff
@@ -0,0 +1,136 @@
+using ePiggyWeb.DataBase;
+using ePiggyWeb.DataManagement.Entries;
+using ePiggyWeb.DataManagement.Goals;
+using ePiggyWeb.Utilities;
+
+namespace ePiggyWeb.DataManagement.MonthlyReport
+{
+    public class MonthlyReportCalculator
+    {
+        private int UserId { get; }
+        private GoalDatabase GoalDatabase { get; }
+        private EntryDatabase EntryDatabase { get; }
+
+        private IEntryList Expenses { get; set; }
+        private IEntryList Income { get; set; }
+        private decimal Balance { get; set; }
+        private decimal AllSavings { get; set; }
+        private IGoalList Goals { get; set; }
+
+        private MonthlyReportResult Result { get; set; }
+
+        private DateTime StartTime { get; set; }
+        private DateTime EndTime { get; set; }
+
+        public MonthlyReportCalculator(GoalDatabase goalDatabase, EntryDatabase entryDatabase, int userId)
+        {
+            GoalDatabase = goalDatabase;
+            EntryDatabase = entryDatabase;
+            UserId = userId;
+        }
+
+        public async Task<MonthlyReportResult> Calculate()
+        { 
+            Result = await GatherDataAsync();
+            CalculateBiggestExpensesCategory();
+            Result.SavedUpGoals = CalculateSavedUpGoals();
+            CalculateGoalReport();
+            return Result;
+        }
+
+        private async Task<MonthlyReportResult> GatherDataAsync()
+        {
+            var today = DateTime.Today;
+            var month = new DateTime(today.Year, today.Month, 1);
+            StartTime = month.AddMonths(-1);
+            EndTime = month.AddDays(-1);
+
+            Expenses = await EntryDatabase.ReadListAsync(UserId, EntryType.Expense);
+            Income = await EntryDatabase.ReadListAsync(UserId, EntryType.Income);
+            Goals = await GoalDatabase.ReadListAsync(UserId);
+
+            var expensesTotal = Expenses.GetSum();
+            var incomeTotal = Income.GetSum();
+            var thisMonthExpenses = Expenses.GetFrom(StartTime).GetTo(EndTime).GetSum();
+            var thisMonthIncome = Income.GetFrom(StartTime).GetTo(EndTime).GetSum();
+            AllSavings = incomeTotal - expensesTotal;
+            Balance = thisMonthIncome - thisMonthExpenses;
+
+            return new MonthlyReportResult(thisMonthExpenses, thisMonthIncome, Balance, StartTime, EndTime);
+
+        }
+
+        private void CalculateBiggestExpensesCategory()
+        {
+            var enumCount = Enum.GetValues(typeof(Importance)).Length;
+            var biggestCategory = (Importance) enumCount;
+            var sum = 0m;
+            foreach (var importance in Enum.GetValues(typeof(Importance)).Cast<Importance>())
+            {
+                var temp = 0M;
+                var expenses = Expenses.GetBy(importance);
+                foreach (var entry in expenses) //not linq since when empty category it crashes
+                {
+                    temp += entry.Amount;
+                }
```

Comment:
When does this actually crash? Maybe these crashes were due to previous division by 0? I've tested myself and didn't find any issues with linq :D

Reply:
For me it happened when any of the categories were empty, but maybe it was coincidence ant other parts crashed :D